File: DailyBrain/rust/src/reviews.rs
// rust/src/reviews.rs
use chrono::NaiveDate;

// ============ Homework Review ============

#[derive(Clone, Debug)]
pub struct HomeworkItem {
    pub id: String,
    pub title: String,
    pub subject: String,
    pub due_date: NaiveDate,
    pub status: String, // "not_started", "in_progress", "completed"
    pub priority: i32,  // 1 = highest
}

pub struct HomeworkContext {
    pub items: Vec<HomeworkItem>,
    pub current_date: NaiveDate,
    pub max_priorities: usize,
}

pub struct HomeworkReviewResult {
    pub overdue: Vec<HomeworkItem>,
    pub due_today: Vec<HomeworkItem>,
    pub upcoming: Vec<HomeworkItem>,
    pub top_priorities: Vec<HomeworkItem>,
    pub note: String,
}

pub struct ReviewHomeworkTask;

impl ReviewHomeworkTask {
    pub fn run(&self, ctx: &HomeworkContext) -> HomeworkReviewResult {
        let incomplete: Vec<_> = ctx
            .items
            .iter()
            .filter(|h| h.status != "completed")
            .cloned()
            .collect();

        let overdue: Vec<_> = incomplete
            .iter()
            .filter(|h| h.due_date < ctx.current_date)
            .cloned()
            .collect();

        let due_today: Vec<_> = incomplete
            .iter()
            .filter(|h| h.due_date == ctx.current_date)
            .cloned()
            .collect();

        let upcoming: Vec<_> = incomplete
            .iter()
            .filter(|h| h.due_date > ctx.current_date)
            .cloned()
            .collect();

        let mut sorted = incomplete.clone();
        sorted.sort_by(|a, b| {
            a.priority
                .cmp(&b.priority)
                .then_with(|| a.due_date.cmp(&b.due_date))
        });
        let top_priorities: Vec<_> = sorted.into_iter().take(ctx.max_priorities).collect();

        let mut notes = Vec::new();
        if !overdue.is_empty() {
            notes.push(format!("⚠️ {} OVERDUE item(s)!", overdue.len()));
        }
        if !due_today.is_empty() {
            notes.push(format!("📅 {} item(s) due TODAY.", due_today.len()));
        }
        if !upcoming.is_empty() {
            notes.push(format!("📋 {} upcoming item(s).", upcoming.len()));
        }

        let note = if notes.is_empty() {
            "✅ All homework complete!".to_string()
        } else {
            notes.join(" ")
        };

        HomeworkReviewResult {
            overdue,
            due_today,
            upcoming,
            top_priorities,
            note,
        }
    }
}

// ============ Bills Review ============

#[derive(Clone, Debug)]
pub struct BillItem {
    pub id: String,
    pub name: String,
    pub amount: f64,
    pub due_date: NaiveDate,
    pub category: String,
    pub is_paid: bool,
    pub is_autopay: bool,
}

pub struct BillsContext {
    pub bills: Vec<BillItem>,
    pub current_date: NaiveDate,
    pub warning_days: i64,
}

pub struct BillsReviewResult {
    pub overdue: Vec<BillItem>,
    pub due_soon: Vec<BillItem>,
    pub upcoming: Vec<BillItem>,
    pub total_due_soon: f64,
    pub note: String,
}

pub struct ReviewBillsTask;

impl ReviewBillsTask {
    pub fn run(&self, ctx: &BillsContext) -> BillsReviewResult {
        let unpaid: Vec<_> = ctx.bills.iter().filter(|b| !b.is_paid).cloned().collect();

        let overdue: Vec<_> = unpaid
            .iter()
            .filter(|b| b.due_date < ctx.current_date)
            .cloned()
            .collect();

        let warning_threshold = ctx.current_date + chrono::Duration::days(ctx.warning_days);
        let due_soon: Vec<_> = unpaid
            .iter()
            .filter(|b| b.due_date >= ctx.current_date && b.due_date <= warning_threshold)
            .cloned()
            .collect();

        let upcoming: Vec<_> = unpaid
            .iter()
            .filter(|b| b.due_date > warning_threshold)
            .cloned()
            .collect();

        let total_due_soon: f64 = overdue.iter().chain(due_soon.iter()).map(|b| b.amount).sum();

        let mut notes = Vec::new();
        if !overdue.is_empty() {
            let total_overdue: f64 = overdue.iter().map(|b| b.amount).sum();
            notes.push(format!(
                "🚨 {} OVERDUE bill(s) totaling ${:.2}!",
                overdue.len(),
                total_overdue
            ));
        }
        if !due_soon.is_empty() {
            let total_soon: f64 = due_soon.iter().map(|b| b.amount).sum();
            notes.push(format!(
                "⏰ {} bill(s) due within {} days (${:.2}).",
                due_soon.len(),
                ctx.warning_days,
                total_soon
            ));
        }

        let note = if notes.is_empty() {
            "✅ All bills are current!".to_string()
        } else {
            notes.join(" ")
        };

        BillsReviewResult {
            overdue,
            due_soon,
            upcoming,
            total_due_soon,
            note,
        }
    }
}

// ============ Todos Review ============

#[derive(Clone, Debug)]
pub struct TodoItem {
    pub id: String,
    pub title: String,
    pub due_date: Option<NaiveDate>,
    pub created_date: NaiveDate,
    pub status: String,   // "pending", "in_progress", "completed", "cancelled"
    pub priority: String, // "low", "medium", "high", "urgent"
    pub context: String,  // "home", "work", "errands", "personal"
}

pub struct TodosContext {
    pub todos: Vec<TodoItem>,
    pub current_date: NaiveDate,
    pub filter_context: Option<String>,
    pub max_display: usize,
}

pub struct TodosReviewResult {
    pub urgent: Vec<TodoItem>,
    pub today: Vec<TodoItem>,
    pub overdue: Vec<TodoItem>,
    pub next_actions: Vec<TodoItem>,
    pub note: String,
}

pub struct ReviewTodosTask;

impl ReviewTodosTask {
    fn priority_order(priority: &str) -> i32 {
        match priority {
            "urgent" => 0,
            "high" => 1,
            "medium" => 2,
            "low" => 3,
            _ => 99,
        }
    }

    pub fn run(&self, ctx: &TodosContext) -> TodosReviewResult {
        let mut active: Vec<_> = ctx
            .todos
            .iter()
            .filter(|t| t.status == "pending" || t.status == "in_progress")
            .cloned()
            .collect();

        if let Some(ref filter) = ctx.filter_context {
            active.retain(|t| &t.context == filter);
        }

        let urgent: Vec<_> = active
            .iter()
            .filter(|t| t.priority == "urgent")
            .cloned()
            .collect();

        let overdue: Vec<_> = active
            .iter()
            .filter(|t| t.due_date.map_or(false, |d| d < ctx.current_date))
            .cloned()
            .collect();

        let today: Vec<_> = active
            .iter()
            .filter(|t| t.due_date == Some(ctx.current_date))
            .cloned()
            .collect();

        let mut sorted = active.clone();
        sorted.sort_by(|a, b| {
            Self::priority_order(&a.priority)
                .cmp(&Self::priority_order(&b.priority))
                .then_with(|| a.created_date.cmp(&b.created_date))
        });
        let next_actions: Vec<_> = sorted.into_iter().take(ctx.max_display).collect();

        let mut notes = Vec::new();
        if !urgent.is_empty() {
            notes.push(format!("🔴 {} URGENT item(s)!", urgent.len()));
        }
        if !overdue.is_empty() {
            notes.push(format!("⚠️ {} overdue item(s).", overdue.len()));
        }
        if !today.is_empty() {
            notes.push(format!("📅 {} item(s) due today.", today.len()));
        }
        notes.push(format!("📋 {} total active items.", active.len()));

        TodosReviewResult {
            urgent,
            today,
            overdue,
            next_actions,
            note: notes.join(" "),
        }
    }
}

File: DailyBrain/cpp/src/reviews.cpp
// cpp/src/reviews.cpp
#include "reviews.h"
#include <algorithm>
#include <sstream>
#include <iomanip>
#include <ctime>

// ============ Homework Review ============

bool ReviewHomeworkTask::date_less_than(const std::tm& a, const std::tm& b) {
    if (a.tm_year != b.tm_year) return a.tm_year < b.tm_year;
    if (a.tm_mon != b.tm_mon) return a.tm_mon < b.tm_mon;
    return a.tm_mday < b.tm_mday;
}

bool ReviewHomeworkTask::date_equal(const std::tm& a, const std::tm& b) {
    return a.tm_year == b.tm_year && a.tm_mon == b.tm_mon && a.tm_mday == b.tm_mday;
}

HomeworkReviewResult ReviewHomeworkTask::run(const HomeworkContext& ctx) {
    HomeworkReviewResult result;
    
    std::vector<HomeworkItem> incomplete;
    for (const auto& h : ctx.items) {
        if (h.status != "completed") {
            incomplete.push_back(h);
        }
    }
    
    for (const auto& h : incomplete) {
        if (date_less_than(h.due_date, ctx.current_date)) {
            result.overdue.push_back(h);
        } else if (date_equal(h.due_date, ctx.current_date)) {
            result.due_today.push_back(h);
        } else {
            result.upcoming.push_back(h);
        }
    }
    
    // Sort by priority, then due date
    auto sorted = incomplete;
    std::sort(sorted.begin(), sorted.end(), [this](const HomeworkItem& a, const HomeworkItem& b) {
        if (a.priority != b.priority) return a.priority < b.priority;
        return date_less_than(a.due_date, b.due_date);
    });
    
    for (int i = 0; i < std::min(static_cast<int>(sorted.size()), ctx.max_priorities); ++i) {
        result.top_priorities.push_back(sorted[i]);
    }
    
    // Generate note
    std::ostringstream oss;
    if (!result.overdue.empty()) {
        oss << "⚠️ " << result.overdue.size() << " OVERDUE item(s)! ";
    }
    if (!result.due_today.empty()) {
        oss << "📅 " << result.due_today.size() << " item(s) due TODAY. ";
    }
    if (!result.upcoming.empty()) {
        oss << "📋 " << result.upcoming.size() << " upcoming item(s). ";
    }
    
    result.note = oss.str().empty() ? "✅ All homework complete!" : oss.str();
    
    return result;
}

// ============ Bills Review ============

bool ReviewBillsTask::date_less_than(const std::tm& a, const std::tm& b) {
    if (a.tm_year != b.tm_year) return a.tm_year < b.tm_year;
    if (a.tm_mon != b.tm_mon) return a.tm_mon < b.tm_mon;
    return a.tm_mday < b.tm_mday;
}

bool ReviewBillsTask::date_less_equal(const std::tm& a, const std::tm& b) {
    return date_less_than(a, b) || (a.tm_year == b.tm_year && a.tm_mon == b.tm_mon && a.tm_mday == b.tm_mday);
}

std::tm ReviewBillsTask::add_days(const std::tm& date, int days) {
    std::tm result = date;
    result.tm_mday += days;
    std::mktime(&result); // normalize
    return result;
}

BillsReviewResult ReviewBillsTask::run(const BillsContext& ctx) {
    BillsReviewResult result;
    result.total_due_soon = 0.0;
    
    std::vector<BillItem> unpaid;
    for (const auto& b : ctx.bills) {
        if (!b.is_paid) {
            unpaid.push_back(b);
        }
    }
    
    std::tm warning_threshold = add_days(ctx.current_date, ctx.warning_days);
    
    for (const auto& b : unpaid) {
        if (date_less_than(b.due_date, ctx.current_date)) {
            result.overdue.push_back(b);
            result.total_due_soon += b.amount;
        } else if (date_less_equal(b.due_date, warning_threshold)) {
            result.due_soon.push_back(b);
            result.total_due_soon += b.amount;
        } else {
            result.upcoming.push_back(b);
        }
    }
    
    // Generate note
    std::ostringstream oss;
    if (!result.overdue.empty()) {
        double total_overdue = 0.0;
        for (const auto& b : result.overdue) total_overdue += b.amount;
        oss << "🚨 " << result.overdue.size() << " OVERDUE bill(s) totaling $"
            << std::fixed << std::setprecision(2) << total_overdue << "! ";
    }
    if (!result.due_soon.empty()) {
        double total_soon = 0.0;
        for (const auto& b : result.due_soon) total_soon += b.amount;
        oss << "⏰ " << result.due_soon.size() << " bill(s) due within "
            << ctx.warning_days << " days ($"
            << std::fixed << std::setprecision(2) << total_soon << "). ";
    }
    
    result.note = oss.str().empty() ? "✅ All bills are current!" : oss.str();
    
    return result;
}

// ============ Todos Review ============

int ReviewTodosTask::priority_order(const std::string& priority) {
    if (priority == "urgent") return 0;
    if (priority == "high") return 1;
    if (priority == "medium") return 2;
    if (priority == "low") return 3;
    return 99;
}

bool ReviewTodosTask::date_less_than(const std::tm& a, const std::tm& b) {
    if (a.tm_year != b.tm_year) return a.tm_year < b.tm_year;
    if (a.tm_mon != b.tm_mon) return a.tm_mon < b.tm_mon;
    return a.tm_mday < b.tm_mday;
}

bool ReviewTodosTask::date_equal(const std::tm& a, const std::tm& b) {
    return a.tm_year == b.tm_year && a.tm_mon == b.tm_mon && a.tm_mday == b.tm_mday;
}

TodosReviewResult ReviewTodosTask::run(const TodosContext& ctx) {
    TodosReviewResult result;
    
    std::vector<TodoItem> active;
    for (const auto& t : ctx.todos) {
        if (t.status == "pending" || t.status == "in_progress") {
            if (ctx.filter_context.empty() || t.context == ctx.filter_context) {
                active.push_back(t);
            }
        }
    }
    
    for (const auto& t : active) {
        if (t.priority == "urgent") {
            result.urgent.push_back(t);
        }
        if (t.has_due_date) {
            if (date_less_than(t.due_date, ctx.current_date)) {
                result.overdue.push_back(t);
            } else if (date_equal(t.due_date, ctx.current_date)) {
                result.today.push_back(t);
            }
        }
    }
    
    // Sort by priority, then created date
    auto sorted = active;
    std::sort(sorted.begin(), sorted.end(), [this](const TodoItem& a, const TodoItem& b) {
        int pa = priority_order(a.priority);
        int pb = priority_order(b.priority);
        if (pa != pb) return pa < pb;
        return date_less_than(a.created_date, b.created_date);
    });
    
    for (int i = 0; i < std::min(static_cast<int>(sorted.size()), ctx.max_display); ++i) {
        result.next_actions.push_back(sorted[i]);
    }
    
    // Generate note
    std::ostringstream oss;
    if (!result.urgent.empty()) {
        oss << "🔴 " << result.urgent.size() << " URGENT item(s)! ";
    }
    if (!result.overdue.empty()) {
        oss << "⚠️ " << result.overdue.size() << " overdue item(s). ";
    }
    if (!result.today.empty()) {
        oss << "📅 " << result.today.size() << " item(s) due today. ";
    }
    oss << "📋 " << active.size() << " total active items.";
    
    result.note = oss.str();
    
    return result;
}
